### PR TITLE
feat(utxo-lib): add full 64 bit support w/ bigints (needed for doge)

### PR DESCRIPTION
**Pull request by williamqin-bitgo**
Depends on PR https://github.com/BitGo/bitcoinjs-lib/pull/36

CLOSES TICKET [BG-49864](https://bitgoinc.atlassian.net/browse/BG-49864), [BG-53245](https://bitgoinc.atlassian.net/browse/BG-53245)

BREAKING CHANGE: UtxoTransaction.fromBuffer interface
- new parameter `amountType` inserted after `noStrict`

This breaking change is small and the function is used primarily internally; users should be using `transaction.ts` `createTransactionFromBuffer` instead so it is unlikely to affect them.

### CODE REVIEW

**Comment by williamqin-bitgo** [comment]
Tests have passed but are flaky, e.g. `yarn install` failing, a `bls-dkg` test timing out (passes when I tried locally)

**Comment by williamqin-bitgo** [comment]
First commit is the feature; all subsequent commits add tests

File: modules/utxo-lib/src/bitgo/Unspent.ts [hunk 1]
```diff
@@ -97,11 +101,27 @@ export function toPrevOutput(u: Unspent, network: Network): PrevOutput {
...
-export function unspentSum(unspents: Unspent[]): number {
-  return unspents.reduce((sum, u) => sum + u.value, 0);
+export function unspentSum<TNumber extends number | bigint>(
+  unspents: Unspent<TNumber>[],
+  sumType: 'inherit' | 'number' | 'bigint' = 'inherit'
+): TNumber {
+  const total = unspents.reduce((sum, u) => sum + BigInt(u.value), BigInt(0));
+  if (sumType === 'bigint' || (unspents.length > 0 && typeof unspents[0].value === 'bigint' && sumType === 'inherit')) {
+    return total as TNumber;
+  } else {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/Unspent.ts:120]: 1. An empty unspents argument with `inherit` will cause the result `0` (`number`). I don't think that is what we want. We should error instead.
2. We are not checking if all unspents values have the same type.

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts [hunk 2]
```diff
@@ -8,29 +8,36 @@ export function varSliceSize(slice: Buffer): number {
...
-    prevOutput?: bitcoinjs.TxOutput[]
-  ): UtxoTransaction {
+    prevOutput?: bitcoinjs.TxOutput<TNumber>[]
+  ): UtxoTransaction<TNumber> {
     if (!network) {
       throw new Error(`must provide network`);
     }
-    return new UtxoTransaction(network, bitcoinjs.Transaction.fromBuffer(buf, noStrict));
+    if (amountType !== 'number' && (getMainnet(network) === networks.dash || getMainnet(network) === networks.zcash)) {
+      throw new Error('dash and zcash must use number amount type; bigint amount type is recommended for doge only');
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoTransaction.ts:38]: What about the other coins? (btc, bch, ... )

File: modules/utxo-lib/test/bitgo/signature.ts [hunk 3]
```diff
@@ -186,21 +189,22 @@ function runTestParseScript(
...
+function assertVerifySignatureEquals<TNumber extends number | bigint>(
+  tx: UtxoTransaction<TNumber>,
+  prevOutputs: TxOutput<TNumber>[],
   value: boolean,
   verificationSettings?: {
     publicKey?: Buffer;
     signatureIndex?: number;
-  }
+  },
+  testOutputAmount = defaultTestOutputAmount as TNumber
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/signature.ts:200]: this does not look right - should probably be an argument instead

File: modules/utxo-lib/test/bitgo/signature.ts [hunk 4]
```diff
@@ -256,37 +283,51 @@ function checkSignTransaction(tx: UtxoTransaction, scriptType: ScriptType2Of3, s
...
-    describe(`verifySignature ${getNetworkName(network)} ${scriptType} ${keyName(k1)} ${keyName(k2)}`, function () {
+    describe(`verifySignature
+${getNetworkName(network)} ${scriptType} ${keyName(k1)} ${keyName(k2)} ${amountType}`, function () {
       it(`verifies half-signed`, function () {
-        checkSignTransaction(getHalfSignedTransaction2Of3(fixtureKeys, k1, k2, scriptType, network), scriptType, [k1]);
+        checkSignTransaction(
+          getHalfSignedTransaction2Of3<TNumber>(fixtureKeys, k1, k2, scriptType, network, amountType),
+          scriptType,
+          [k1],
+          toTNumber<TNumber>(defaultTestOutputAmount, amountType)
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/signature.ts:309]: :+1: this is the right way to convert to `amountType`

File: modules/utxo-lib/test/bitgo/signature.ts [hunk 5]
```diff
@@ -302,11 +343,15 @@ describe('Signature (scriptTypes2Of3)', function () {
...
+        runTestCheckSignatureVerify<bigint>(network, scriptType, undefined, undefined, 'bigint');
 
         getSignKeyCombinations(2).map(([k1, k2]) => {
           runTestCheckSignatureVerify(network, scriptType, k1, k2);
+          runTestCheckSignatureVerify<bigint>(network, scriptType, k1, k2, 'bigint');
           runTestCheckScriptStructure(network, scriptType, k1, k2);
+          runTestCheckScriptStructure<bigint>(network, scriptType, k1, k2, 'bigint');
           runTestParseScript(network, scriptType, k1, k2);
+          runTestParseScript<bigint>(network, scriptType, k1, k2, 'bigint');
         });
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/signature.ts:355]: let's add a second outer loop instead

File: modules/utxo-lib/test/bitgo/wallet/WalletUnspent.ts [hunk 6]
```diff
@@ -53,22 +54,36 @@ describe('WalletUnspent', function () {
...
   });
 
-  function runTestSignUnspent(scriptType: outputScripts.ScriptType2Of3, signer: string, cosigner: string) {
-    it(`can be signed [scriptType=${scriptType} signer=${signer} cosigner=${cosigner}]`, function () {
+  function runTestSignUnspent<TNumber extends number | bigint>(
+    scriptType: outputScripts.ScriptType2Of3,
+    signer: string,
+    cosigner: string,
+    amountType: 'number' | 'bigint' = 'number',
+    testOutputAmount: TNumber = defaultTestOutputAmount as TNumber
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/wallet/WalletUnspent.ts:62]: `toTNumber` would be better here

File: modules/utxo-lib/test/bitgo/wallet/util.ts [hunk 7]
```diff
@@ -14,17 +14,17 @@ export function mockOutputId(vout: number): string {
...
 }
 
-export function mockWalletUnspent(
+export function mockWalletUnspent<TNumber extends number | bigint = number>(
   network: Network,
   {
     chain = 0 as ChainCode,
     index = 0,
-    value = defaultTestOutputAmount,
+    value = defaultTestOutputAmount as TNumber,
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/wallet/util.ts:22]: instead of casting, it is probably better to not provide a default argument here at all

File: modules/utxo-lib/test/integration_local_rpc/generate/main.ts [hunk 8]
```diff
@@ -109,16 +113,38 @@ async function createTransactionsForScriptType(
     return;
   }
 
-  const deposit2Txid = await sendFromFaucet(rpc, address, 1);
+  let amount: number | string = 1;
+  switch (protocol.network) {
+    case utxolib.networks.dogecoinTest:
+      amount = 109999998.00000001;
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/integration_local_rpc/generate/main.ts:119]: I assume this is to exercise the `bigint` precision? Let's add a comment!

File: modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts [hunk 9]
```diff
@@ -41,7 +41,7 @@ function getDockerParams(network: Network): DockerImageParams {
     case utxolib.networks.dashTest:
       return dockerImage('dashpay/dashd:0.16.1.1', 'dashd');
     case utxolib.networks.dogecoinTest:
-      return dockerImage('williamqinbitgo/dogeimage:1.14.5-v3', 'dogecoind');
+      return dockerImage('williamqinbitgo/dogeimage:1.14.5-v2', 'dogecoind');
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts:44]: why?

File: modules/utxo-lib/test/integration_local_rpc/parse.ts [hunk 10]
```diff
@@ -89,15 +97,27 @@ function runTestParse(protocol: Protocol, txType: FixtureTxType, scriptType: Scr
...
-      return getPrevOutput(input).value * 1e8;
+    function getValueScaled(value: number, scaleFactor = 1e8): TNumber {
+      if (amountType === 'number') {
+        return (Number(value) * scaleFactor) as TNumber;
+      } else {
+        const integerPart = Math.floor(value);
+        let decimalPart = value - integerPart;
+        decimalPart = Math.round(decimalPart * scaleFactor); // scale and fix floating point
+
+        return (BigInt(integerPart) * BigInt(scaleFactor) + BigInt(decimalPart)) as TNumber;
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/integration_local_rpc/parse.ts:108]: seeing this a second time now - this should be a separate func with own tests

File: modules/utxo-lib/test/transaction_util.ts [hunk 11]
```diff
@@ -66,12 +67,25 @@ export function mockTransactionId(v = 0xff): string {
   return Buffer.alloc(32).fill(v).toString('hex');
 }
 
-export function getPrevOutput(
+export function toTNumber<TNumber extends number | bigint>(v: any | undefined, t: 'number' | 'bigint'): TNumber {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/transaction_util.ts:70]:
```suggestion
export function toTNumber<TNumber extends number | bigint>(v: number | bigint | undefined, t: 'number' | 'bigint'): TNumber | undefined {
```

File: modules/utxo-lib/test/transaction_util.ts [hunk 12]
```diff
@@ -66,12 +67,25 @@ export function mockTransactionId(v = 0xff): string {
...
+    return BigInt(v) as TNumber;
+  }
+  throw new Error();
+}
+
+export function getPrevOutput<TNumber extends number | bigint = number>(
   scriptType: ScriptType2Of3 | 'p2shP2pk',
   vout = 0,
-  value = defaultTestOutputAmount,
+  value = defaultTestOutputAmount as TNumber,
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/transaction_util.ts:86]: let's drop the default value instead

File: modules/utxo-lib/test/transaction_util.ts [hunk 13]
```diff
@@ -85,40 +99,44 @@ export function getPrevOutput(
   };
 }
 
-export function getPrevOutputs(
+export function getPrevOutputs<TNumber extends number | bigint = number>(
   scriptType: ScriptType2Of3 | 'p2shP2pk',
-  value = defaultTestOutputAmount,
+  value = defaultTestOutputAmount as TNumber,
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/transaction_util.ts:104]: drop default value

File: modules/utxo-lib/test/transaction_util.ts [hunk 14]
```diff
@@ -85,40 +99,44 @@ export function getPrevOutput(
...
   prevOutputs.forEach(({ txid, vout }) => {
     txBuilder.addInput(txid, vout);
   });
 
   const recipientScript = createScriptPubKey(fixtureKeys, 'p2pkh', networks.bitcoin);
-  txBuilder.addOutput(recipientScript, outputAmount - 1000);
+  const tNumOutput = (
+    amountType === 'number' ? Number(outputAmount) - 1000 : BigInt(outputAmount) - BigInt(1000)
+  ) as TNumber;
+  txBuilder.addOutput(recipientScript, tNumOutput);
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/transaction_util.ts:139]:
```suggestion
  txBuilder.addOutput(recipientScript, toTNumber(BigInt(outputAmount) - 1000n, amountType));
```

File: modules/utxo-lib/test/bitgo/wallet/WalletUnspent.ts [hunk 15]
```diff
@@ -53,22 +54,36 @@ describe('WalletUnspent', function () {
...
       ];
-      const txb = createTransactionBuilderForNetwork(network);
-      txb.addOutput(getWalletAddress(walletKeys, 0, 100, network), unspentSum(unspents) - 100);
+      const txb = createTransactionBuilderForNetwork<TNumber>(network);
+      let outputValue = unspentSum(unspents) as TNumber;
+      if (amountType === 'number') {
+        outputValue = ((outputValue as number) - 100) as TNumber;
+      } else {
+        outputValue = ((outputValue as bigint) - BigInt(100)) as TNumber;
+      }
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/test/bitgo/wallet/WalletUnspent.ts:85]: use `toTNumber` instead

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts (hunk 9 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts:44]: It turns out that the `v3` version I published is actually broken, and that the regtest fixtures had been generated incorrectly (from an earlier PR with one of my commits). This commit should fix the generation.

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts (hunk 2 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/UtxoTransaction.ts:38]: Strictly speaking, `bigint` _should_ work with the other coins (it just definitely doesn't work with zcash/dash).
Also, some tests use `bigint` with BTC so I only restricted dash and zcash to accommodate those tests.

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 1 above)
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/Unspent.ts:120]: consider https://github.com/BitGo/BitGoJS/commit/165cbd92b11bace5dbe5ad0717a61649ea2b4dce

File: modules/utxo-lib/src/bitgo/transaction.ts [hunk 16]
```diff
@@ -100,27 +112,35 @@ export function createTransactionBuilderForNetwork(
...
+        undefined,
+        prevOutputs as TxOutput[]
+      ) as unknown as UtxoTransactionBuilder<TNumber>;
     case networks.zcash:
-      return ZcashTransactionBuilder.fromTransaction(tx as ZcashTransaction, undefined, prevOutputs);
+      return ZcashTransactionBuilder.fromTransactionZcash(
+        tx as unknown as ZcashTransaction,
+        undefined,
+        prevOutputs as TxOutput[]
+      ) as unknown as UtxoTransactionBuilder<TNumber>;
```
[the remark this answers is not in the document]
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:143]: To make the implementation cleaner, I had `ZcashTransaction` and `DashTransaction` (and respective txbs) extend the `number` **only** version of `UtxoTransaction(Builder)`. 
However, this makes it impossible to extend the `fromBuffer` or `fromTransaction` methods on `UtxoTransaction(Builder)` so I had to rename them.

File: modules/utxo-lib/test/bitgo/signature.ts (hunk 3 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/bitgo/signature.ts:200]: removing default value

File: modules/utxo-lib/test/bitgo/signature.ts [hunk 17]
```diff
@@ -210,15 +214,26 @@ function assertVerifySignatureEquals(
   });
 }
 
-function checkSignTransaction(tx: UtxoTransaction, scriptType: ScriptType2Of3, signKeys: bip32.BIP32Interface[]) {
-  const prevOutputs = getPrevOutputs(scriptType, defaultTestOutputAmount) as TxOutput[];
+function checkSignTransaction<TNumber extends number | bigint>(
+  tx: UtxoTransaction<TNumber>,
+  scriptType: ScriptType2Of3,
+  signKeys: bip32.BIP32Interface[],
+  testOutputAmount = defaultTestOutputAmount as TNumber
```
[the remark this answers is not in the document]
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/bitgo/signature.ts:221]: removing default value

File: modules/utxo-lib/test/bitgo/wallet/util.ts (hunk 7 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/bitgo/wallet/util.ts:22]: moved value out of curly braces, turned into a regular arg

File: modules/utxo-lib/test/transaction_util.ts (hunk 11 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/transaction_util.ts:70]: I had copied the similar function in `bitcoinjs-lib`, but it turns out that v shouldn't ever be `undefined`.

File: modules/utxo-lib/test/integration_local_rpc/parse.ts (hunk 10 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/integration_local_rpc/parse.ts:108]: created new function in `transaction_util`

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 1 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts:120]: I removed the 'inherit' option and defaulted to 'number', since an empty unspents argument would previously result in `0` (`number`) but that doesn't make sense anymore with the `inherit` sum type.

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:143]: I think we can, but it will expand the size of this PR by quite a bit (I think there are several files to change)

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 1 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts:120]: I also used type assertions, and so if the unspents array mixes `bigint` and `number` there will be a runtime error

Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts]: I noticed that in https://github.com/BitGo/BitGoJS/commit/165cbd92b11bace5dbe5ad0717a61649ea2b4dce `toTNumber` was put in `src` code (instead of `test`) and exported - is that worth doing with `toTNumber` and `getValueScaled` (two helper functions I added to `transaction_util.ts`)

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/transaction.ts:143]: OK let's leave it as is then

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 1 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts]: I ended up moving `toTNumber` and `getValueScaled` to `src`

File: modules/utxo-lib/src/bitgo/tnumber.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  throw new Error();
+}
+
+/** Scale a decimal amount (optionally to a bigint) */
+export function getValueScaled<TNumber extends number | bigint>(
+  value: number,
+  amountType: 'number' | 'bigint' = 'number',
+  scaleFactor = 1e8
+): TNumber {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/tnumber.ts:23]: please add a test for this (could be a separate PR)

**Review by OttoAllmendinger — DISMISSED** [review]

File: modules/utxo-lib/src/bitgo/tnumber.ts (hunk 18 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/tnumber.ts]: Added

**Comment by williamqin-bitgo** [comment]
Tests seem to be more consistent after rebasing

File: modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts (hunk 9 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts]: I updated the v3 docker image to be working and roughly 3 GB smaller in size so there's no need to revert anymore

Comment (OttoAllmendinger) [reply modules/utxo-lib/test/integration_local_rpc/generate/regtestNode.ts]: ideally you would have called it `-v4` then ;)

**Review by OttoAllmendinger — DISMISSED** [review]

File: modules/utxo-lib/src/bitgo/Unspent.ts [hunk 19]
```diff
@@ -97,11 +101,33 @@ export function toPrevOutput(u: Unspent, network: Network): PrevOutput {
...
+/**
+ * Sum the values of the unspents.
+ * Throws error if sum is not a safe integer value, or if unspent amount types do not match `amountType`
+ * @param unspents - array of unspents to sum
+ * @param amountType - expected value type of unspents
+ * @return unspentSum - type matches amountType
+ */
+export function unspentSum<TNumber extends number | bigint>(
+  unspents: Unspent<TNumber>[],
+  amountType: 'number' | 'bigint' = 'number'
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/Unspent.ts:122]: New required field in parameters. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts [hunk 20]
```diff
@@ -8,29 +8,36 @@ export function varSliceSize(slice: Buffer): number {
...
     this.locktime = transaction.locktime;
     this.ins = transaction.ins.map((v) => ({ ...v }));
     this.outs = transaction.outs.map((v) => ({ ...v }));
   }
 
-  static fromBuffer(
+  static fromBuffer<TNumber extends number | bigint = number>(
     buf: Buffer,
     noStrict: boolean,
+    amountType: 'number' | 'bigint' = 'number',
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/UtxoTransaction.ts:30]: New required field in parameters. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/dash/DashTransaction.ts [hunk 21]
```diff
@@ -40,8 +40,8 @@ export class DashTransaction extends UtxoTransaction {
     return new DashTransaction(tx.network, tx);
   }
 
-  static fromBuffer(buffer: Buffer, noStrict: boolean, network: Network): DashTransaction {
-    const baseTx = UtxoTransaction.fromBuffer(buffer, true, network);
+  static fromBufferDash(buffer: Buffer, noStrict: boolean, network: Network): DashTransaction {
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/dash/DashTransaction.ts:43]: Method removal. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/dash/DashTransactionBuilder.ts [hunk 22]
```diff
@@ -21,7 +21,7 @@ export class DashTransactionBuilder extends UtxoTransactionBuilder<DashTransacti
     this.tx.extraPayload = extraPayload;
   }
 
-  static fromTransaction(
+  static fromTransactionDash(
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/dash/DashTransactionBuilder.ts:24]: Method removal. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/transaction.ts [hunk 23]
```diff
@@ -9,32 +9,44 @@ import { DashTransactionBuilder } from './dash/DashTransactionBuilder';
 import { ZcashTransactionBuilder } from './zcash/ZcashTransactionBuilder';
 import { ZcashNetwork, ZcashTransaction } from './zcash/ZcashTransaction';
 
-export function createTransactionFromBuffer(
+export function createTransactionFromBuffer<TNumber extends number | bigint = number>(
   buf: Buffer,
   network: Network,
+  amountType: 'number' | 'bigint' = 'number',
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/transaction.ts:15]: Change in required parameters. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/transaction.ts [hunk 24]
```diff
@@ -9,32 +9,44 @@ import { DashTransactionBuilder } from './dash/DashTransactionBuilder';
...
   throw new Error(`invalid network`);
 }
 
 /* istanbul ignore next */
-export function createTransactionFromHex(hex: string, network: Network): UtxoTransaction {
-  return createTransactionFromBuffer(Buffer.from(hex, 'hex'), network);
+export function createTransactionFromHex<TNumber extends number | bigint = number>(
+  hex: string,
+  network: Network,
+  amountType: 'number' | 'bigint' = 'number'
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/transaction.ts:47]: Change in required parameters. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/transaction.ts [hunk 25]
```diff
@@ -65,7 +77,7 @@ export function setTransactionBuilderDefaults(
       txb.setVersion(version);
       break;
     case networks.zcash:
-      (txb as ZcashTransactionBuilder).setDefaultsForVersion(network, version);
+      (txb as unknown as ZcashTransactionBuilder).setDefaultsForVersion(network, version);
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/transaction.ts:80]: External users complain about the use of `unknown` and `any`. Is it possible to identify this?

File: modules/utxo-lib/src/bitgo/zcash/ZcashTransaction.ts [hunk 26]
```diff
@@ -94,7 +94,7 @@ export class ZcashTransaction extends UtxoTransaction {
     this.consensusBranchId = consensusBranchId ?? getDefaultConsensusBranchIdForVersion(network, this.version);
   }
 
-  static fromBuffer(buffer: Buffer, __noStrict: boolean, network?: ZcashNetwork): ZcashTransaction {
+  static fromBufferZcash(buffer: Buffer, __noStrict: boolean, network?: ZcashNetwork): ZcashTransaction {
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/zcash/ZcashTransaction.ts:97]: Method removal. Please record as BREAKING CHANGE in commits.

File: modules/utxo-lib/src/bitgo/zcash/ZcashTransactionBuilder.ts [hunk 27]
```diff
@@ -21,7 +21,7 @@ export class ZcashTransactionBuilder extends UtxoTransactionBuilder<ZcashTransac
     return new ZcashTransaction(network as ZcashNetwork, tx as ZcashTransaction);
   }
 
-  static fromTransaction(
+  static fromTransactionZcash(
```
Comment (mmcshinsky-bitgo) [inline modules/utxo-lib/src/bitgo/zcash/ZcashTransactionBuilder.ts:24]: Method removal. Please record as BREAKING CHANGE in commits.

**Review by mmcshinsky-bitgo — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 19 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts:122]: Isn't this optional? It has a default value.

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts (hunk 20 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/UtxoTransaction.ts:30]: ^^

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 23 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:15]: See first conversation

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 24 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:47]: See first conversation

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 19 above)
Comment (mmcshinsky-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts:122]: Ope, yup. Ignore those ones

File: modules/utxo-lib/src/bitgo/dash/DashTransaction.ts (hunk 21 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/dash/DashTransaction.ts:43]: Looks like this is a problem with trying to have `Dash` and `Zcash` transaction only inherit `UtxoTransaction<number>`, implementing `TNumber` for them instead (see https://github.com/BitGo/BitGoJS/pull/2533#discussion_r928551322)

File: modules/utxo-lib/src/bitgo/Unspent.ts (hunk 19 above)
Comment (mmcshinsky-bitgo) [reply modules/utxo-lib/src/bitgo/Unspent.ts:122]: Ope, yup. Ignore those.

File: modules/utxo-lib/src/bitgo/dash/DashTransaction.ts (hunk 21 above)
Comment (mmcshinsky-bitgo) [reply modules/utxo-lib/src/bitgo/dash/DashTransaction.ts:43]: We'll still need it recorded as a breaking change or refactor it.

Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/dash/DashTransaction.ts:43]: Yep, refactoring

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts (hunk 20 above)
Comment (williamqin-bitgo) [inline modules/utxo-lib/src/bitgo/UtxoTransaction.ts:30]: I noticed here that I changed the order of the arguments though, going to reorder so that `amountType` is at the end

File: modules/utxo-lib/src/bitgo/dash/DashTransactionBuilder.ts (hunk 22 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/dash/DashTransactionBuilder.ts:24]: 👍  refactoring

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:143]: Looks like I should do this in this PR so we don't get breaking changes

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 25 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/transaction.ts:80]: I think I can remove this by refactoring

File: modules/utxo-lib/src/bitgo/zcash/ZcashTransaction.ts (hunk 26 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/zcash/ZcashTransaction.ts:97]: refactoring

File: modules/utxo-lib/src/bitgo/zcash/ZcashTransactionBuilder.ts (hunk 27 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/zcash/ZcashTransactionBuilder.ts:24]: refactoring

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts (hunk 20 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/UtxoTransaction.ts:30]: Oh, this isn't actually possible because `bitcoinjs-lib` has the third argument as `amountType`

File: modules/utxo-lib/src/bitgo/transaction.ts (hunk 23 above)
Comment (williamqin-bitgo) [inline modules/utxo-lib/src/bitgo/transaction.ts:15]: This can be reordered though, reordering

File: modules/utxo-lib/src/bitgo/UtxoTransaction.ts (hunk 20 above)
Comment (williamqin-bitgo) [reply modules/utxo-lib/src/bitgo/UtxoTransaction.ts:30]: I think this is an acceptable breaking change, going to record in commit message

**Comment by williamqin-bitgo** [comment]
Breaking change: `UtxoTransaction.fromBuffer` interface changed, new parameter `amountType` inserted after `noStrict`

**Review by mmcshinsky-bitgo — APPROVED** [review]
